File: CSharp/src/FAI.Router/JudgeLogic/Judge.cs
using AI.DataStructs.Algebraic;
using FAI.Router.RotationTracking;

namespace FAI.Router.JudgeLogic;


/// <summary>
/// Судья, автоматическая оценка качества решения.
/// Замером факта судья не занимается: это работа SpecOutputService, и обе спецификации
/// приходят к судье готовыми. Иначе зона оценки зависела бы от зоны сервисов, а та уже
/// зависит от нее, из-за чего получилась бы кольцевая связь.
/// </summary>
public class Judge
{
    /// <summary>
    /// Матрица трансформации (обучаемая)
    /// </summary>
    public Matrix TransformerW { get; set; } = Matrix.Identity(Settings.FeaturesSpecDim);


    /// <summary>
    /// Оценка качества: близость факта к заказу, пропущенному через обучаемую матрицу
    /// </summary>
    /// <param name="inputSpec">Запрашиваемые параметры (ТЗ)</param>
    /// <param name="actualSpec">Фактические параметры ответа</param>
    public double GetScore(Specifications inputSpec, Specifications actualSpec)
    {
        Vector transformed = TransformerW.MulMatrOnVectColumn(inputSpec.FeaturesSpecificationVector);

        return actualSpec.FeaturesSpecificationVector.Cos(transformed);
    }

    /// <summary>
    /// Проставляет оценку хода в его трассировку и тем замыкает круг: ход состоялся,
    /// судья его оценил, и трассировка стала обучающим примером
    /// </summary>
    /// <param name="trace">Трассировка хода</param>
    /// <param name="inputSpec">Запрашиваемые параметры (ТЗ)</param>
    /// <param name="actualSpec">Фактические параметры ответа</param>
    public double Rate(Tracert trace, Specifications inputSpec, Specifications actualSpec)
    {
        trace.Score = GetScore(inputSpec, actualSpec);

        return trace.Score;
    }

    /// <summary>
    /// Режим критика: расхождения между ТЗ и фактом по каждому пункту.
    /// В отличие от GetScore сравнивает пункты по отдельности, без вектора и матрицы,
    /// поэтому объяснимо для человека и не зависит от масштаба координат.
    /// </summary>
    /// <param name="inputSpec">Запрашиваемые параметры</param>
    /// <param name="actualSpec">Фактические параметры ответа</param>
    /// <param name="content">Оценка содержания; с ней разбор включает и расхождения по содержанию</param>
    public static DiffSpec Criticize(Specifications inputSpec, Specifications actualSpec, ContentReview? content = null) =>
        DiffSpec.Compare(inputSpec, actualSpec, content);

    /// <summary>
    /// Итоговая оценка ответа: содержание и форма с долями из <see cref="Settings.ContentWeight"/>.
    /// Форма здесь это доля выполненных пунктов формы критика: число объяснимое и не зависящее от
    /// обучаемой матрицы. Без оценки содержания итог равен форме.
    /// </summary>
    /// <param name="critic">Разбор критика</param>
    /// <param name="content">Оценка содержания; пусто, если ее не делали</param>
    public static double Assess(DiffSpec critic, ContentReview? content)
    {
        double formScore = 1 - critic.FormDeviation;

        return content is null ? formScore : Settings.ContentWeight * content.Score + (1 - Settings.ContentWeight) * formScore;
    }

    /// <summary>
    /// Отчет: оценки содержания, формы и итог, затем все проваленные пункты ТЗ и замечания судьи
    /// </summary>
    /// <param name="critic">Разбор критика</param>
    /// <param name="content">Оценка содержания; пусто, если ее не делали</param>
    public static string Report(DiffSpec critic, ContentReview? content)
    {
        string form = Format(1 - critic.FormDeviation);
        string head = content is null
            ? $"Форма {form}"
            : $"Содержание {Format(content.Score)}, форма {form}, итог {Format(Assess(critic, content))}";

        IEnumerable<string> lines =
        [
            head,
            .. critic.Mismatches.Select(item => $"{item.Field}: заказано {item.Requested}, получено {item.Actual}"),
            .. (content?.Issues ?? []).Select(issue => "- " + issue),
        ];

        return string.Join(Environment.NewLine, lines);
    }

    private static string Format(double value) => value.ToString("0.00", System.Globalization.CultureInfo.InvariantCulture);
}

File: Python/fai_router/judge.py
from __future__ import annotations

import numpy as np

from fai_router.content_review import ContentReview
from fai_router.diff_spec import DiffSpec
from fai_router.settings import Settings
from fai_router.specifications import Specifications
from fai_router.tracking import Tracert


def cosine(a: np.ndarray, b: np.ndarray, eps: float = 1e-12) -> float:
    return float(a @ b / (np.linalg.norm(a) * np.linalg.norm(b) + eps))


class Judge:
    """Судья, автоматическая оценка качества решения. Замером факта судья не занимается:
    обе спецификации приходят к нему готовыми."""

    def __init__(self):
        # Матрица трансформации (обучаемая). Единичная на старте означает наивное допущение,
        # будто факт обязан совпасть с заказом
        self.transformer_w: np.ndarray = np.eye(Settings.features_spec_dim())

    def get_score(self, input_spec: Specifications, actual_spec: Specifications) -> float:
        """Близость факта к заказу, пропущенному через обучаемую матрицу."""
        transformed = self.transformer_w @ input_spec.feature_vector()
        return cosine(actual_spec.feature_vector(), transformed)

    def rate(self, trace: Tracert, input_spec: Specifications, actual_spec: Specifications) -> float:
        """Проставляет оценку хода в его трассировку и тем замыкает круг: ход состоялся,
        судья его оценил, и трассировка стала обучающим примером."""
        trace.score = self.get_score(input_spec, actual_spec)
        return trace.score

    @staticmethod
    def criticize(input_spec: Specifications, actual_spec: Specifications,
                  content: ContentReview | None = None) -> DiffSpec:
        """Режим критика: расхождения между заданием и фактом по каждому пункту; с оценкой
        содержания и по содержанию."""
        return DiffSpec.compare(input_spec, actual_spec, content)

    @staticmethod
    def assess(critic: DiffSpec, content: ContentReview | None) -> float:
        """Итоговая оценка ответа: содержание и форма с долями из Settings.content_weight. Форма
        здесь это доля выполненных пунктов формы критика: число объяснимое и не зависящее от
        обучаемой матрицы. Без оценки содержания итог равен форме."""
        form_score = 1.0 - critic.form_deviation
        if content is None:
            return form_score
        return Settings.content_weight * content.score + (1.0 - Settings.content_weight) * form_score

    @staticmethod
    def report(critic: DiffSpec, content: ContentReview | None) -> str:
        """Отчет: оценки содержания, формы и итог, затем все проваленные пункты задания и
        замечания судьи."""
        form = f"{1.0 - critic.form_deviation:.2f}"
        head = (f"Форма {form}" if content is None
                else f"Содержание {content.score:.2f}, форма {form}, итог {Judge.assess(critic, content):.2f}")
        lines = [head]
        lines += [f"{item.field}: заказано {item.requested}, получено {item.actual}" for item in critic.mismatches]
        lines += [f"- {issue}" for issue in (content.issues if content is not None else [])]
        return "\n".join(lines)
